File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque


def solution(n, wires):
    answer = n
    
    graph = [[False for _ in range(n+1)] for __ in range(n+1)]
    # 그래프 선 연결
    for u, v in wires:
        graph[u][v] = True
        graph[v][u] = True
    
    for u, v in wires:
        graph[u][v] = False
        graph[v][u] = False
        
        answer = min(answer, search(n, graph))
        
        graph[u][v] = True
        graph[v][u] = True
    
    return answer


def search(n, graph):
    visited = [False for _ in range(n+1)]
    results = []
    for i in range(1, n+1):
        if visited[i]:
            continue
        result = bfs(i, graph, visited)
        results.append(result)
        
    return abs(results[0] - results[1])

def bfs(vertex, graph, visited):
    n = len(graph) - 1
    visited[vertex] = True
    answer = 1
    
    dq = deque()
    dq.append(vertex)
    
    while dq:
        cur_vertex = dq.pop()
        for i in range(1, n+1):
            if i == cur_vertex or not graph[cur_vertex][i] or visited[i]:
                continue
            dq.append(i)
            visited[i] = True
            answer += 1
    
    return answer
    
    

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        HashSet<Integer>[] graph = new HashSet[n + 1];
        for(int i = 0; i <= n; i++){
            graph[i] = new HashSet<>();        
        }
        
        int answer = Integer.MAX_VALUE;
        
        for(int i = 0; i < wires.length; i++){
            int u = wires[i][0];
            int v = wires[i][1];
            
            graph[u].add(v);
            graph[v].add(u);
        }
        
        for(int i = 0; i < wires.length; i++){
            int u = wires[i][0];
            int v = wires[i][1];
            
            graph[u].remove(v);
            graph[v].remove(u);
            
            int checkResult = check(n, graph);
            answer = Math.min(answer, checkResult);
            
            graph[u].add(v);
            graph[v].add(u);
        }
        
        
        return answer;
    }
    
    public int check(int n, HashSet<Integer>[] graph){        
        boolean[] visited = new boolean[n + 1];
        ArrayList<Integer> grids = new ArrayList<>();
        
        for(int i = 1; i <= n; i++){
            if(visited[i]) continue;
            int result = bfs(n, i, graph, visited);
            grids.add(result);
        }
        
        return (int) Math.abs(grids.get(0) - grids.get(1));
    }
    
    public int bfs(int n, int start, HashSet<Integer>[] graph, boolean[] visited){
        Deque<Integer> deque = new ArrayDeque<>();
        deque.add(start);
        visited[start] = true;
        int tops = 1;
        
        while(!deque.isEmpty()){
            int cur = deque.poll();
            
            for(int next: graph[cur]){
                if(visited[next]) continue;
                deque.add(next);
                visited[next] = true;
                tops++;
            }
        }
        
        return tops;
    }
}